### DISCUSSION TITLE
FlorisBoard 0.3.16 input logic and editor caching major rework

### DISCUSSION THREAD

**Discussion by patrickgold** [florisboard/florisboard#1827] [General]
This discussion serves as a collection for all feedback/issues surrounding the new input connection logic introduced in #1822 and [0.3.16-beta01](https://github.com/florisboard/florisboard/releases/tag/v0.3.16-beta01).

Please post your feedback here if it concerns the following:
- General feedback to the overhaul: Is input now faster/laggier/the same as before? Did the input experience change for the better/the worse/stayed the same?
- Input into specific or any apps does not work correctly or at all
- Input randomly overwrites stuff
- Selection/cursor movement is messed up
- Delete character/words does not work
- Other issues in this area

If you have an issue in this area which is new since 0.3.15, please make sure to post it here (also specify the app and Android version you are on) and I will try to fix them in the following beta releases.

Thanks a lot!

See PR #1832, which addresses feedback gathered from this thread and other issues

**Comment by pokegamer5547** [comment]
great work! So far everything is working fine and speed was improved a lot.  expect one tiny insignificant bug i noticed.
swiping left/right on space doesn't scroll the cursor in termux which was working in previous version.

**Reply by patrickgold** [reply]
Thanks for the feedback! You are correct, for the left/right swipe I forgot to check if the editor is a raw or rich one and I always assume it is rich. Will fix this for the next beta.

**Reply by patrickgold** [reply]
Commit https://github.com/florisboard/florisboard/pull/1832/commits/5bf4819d83617182c912ab89139bb1e9985e8d17 of PR #1832 fixes this issue, for raw input editors I now send repeated arrow keys, for rich input editors I set the selection via the API. Will be parts of the next beta release 0.3.16-beta02, thanks again for reporting!

**Reply by pokegamer5547** [reply]
oh i understand, you're welcome! I'll be sure to test it once it becomes available. :)

**Reply by pokegamer5547** [reply]
i have tested the change in termux just now and the swiping left/right bug has been fixed! But there's one small thing i'm curios about. 
 You see, after typing something something in termux, if i hold down the shift key while sliding on space to move the cursor, for some reason tbe text i typed was getting deleted. It's not a big deal but i'm curious to know why it happened.  
Edit: i just tested an older version (0.3.15 (80)) from fdroid and it has same problem. So it was probably not because of input change i guess.

**Reply by patrickgold** [reply]
Glad to hear it is fixed for you too!

As for the shift + space swipe: Technically you just do repeated shift+arrow left/right taps with the space swipe, and shift + arrow keys don't really work on Unix-like terminals, not even on desktop (here it just inputs a character). So yeah not really a FlorisBoard issue. What I could do though is to ignore the shift key while space swiping, but I don't know it that's a good thing, as some raw editors may support shift+arrow keys.

**Reply by pokegamer5547** [reply]
Ah i see, that's interesting. if it's not a problem by florisboard and it only happens for terminals then i dont think it's such a big issue. 
Btw idk how i didnt noticed this before (man i feel little awkward😅) but, in current beta 2 version, shift+space swipe isnt selecting the text  not just termux, but in any other app as well. Is it a known bug?

**Reply by pokegamer5547** [reply]
hello, i'm sorry to bother you i just thought maybe you didnt noticed so i'm replying. The bug of not being anle to select text using shift+swipe is still isn't working on any app for me using latest beta 2.
Sorry again ik maybe you are just busy but i just wanted to inform.

**Reply by patrickgold** [reply]
Yes I am aware, didn't have too much time fixing this bug yet, I am sorry. It is tracked in #1843 though, so if you follow that issue you'll get noticed once I fix it.

**Reply by pokegamer5547** [reply]
oh i understand it's alright! I was just thought to remind since i didnt got any response and i thought you didnt noticed. It's not serious so take your time. :)

**Comment by Humorist2601** [comment]
Another bug that I encountered, when you glide type a word, it stays permanently on the Smartbar. Even when you start typing normally.

Edit: Shift+Spacebar not longer selecting text atleast for me

**Reply by patrickgold** [reply]
> Another bug that I encountered, when you glide type a word, it stays permanently on the Smartbar. Even when you start typing normally.

Correct, this seems to be because the update state logic does not clear the word at all due to an incomplete logic. Will address this in the follow up beta.

> Edit: Shift+Spacebar not longer selecting text atleast for me

Can also reproduce this bug, is a bit more tricky to fix but hope I can also fix this in the next beta

**Reply by Humorist2601** [reply]
Sorry keep bothering you with bugs :( but I have encountered another one.
1. Set Space bar Long press gesture to open Clipboard Manager/History
2. Long press Space bar (Nothing happens)
3. Press any other key, Clipboard Manager Opens

Edit: Fixed in #1832

**Reply by patrickgold** [reply]
You are indeed right, this seems to be a state update bug. Is weird though because the clipboard button in the smartbar (which uses the same state logic) works correctly.

**Reply by patrickgold** [reply]
PR #1832 fixes the space bar long press for the clipboard manager action, the reason was that the action ran in a different thread than main, which confused the logic.

**Reply by patrickgold** [reply]
Commit https://github.com/florisboard/florisboard/pull/1832/commits/60221743f6a7721bb006cc946fcf19ddfb308d5c in PR #1832 fixes the Smartbar UI update bug, so the glided word disappears again after you release your finger. Thanks for reporting!

**Comment by papjul** [comment]
Regression report here: Deleting characters in Markor app in TODO mode has become terrible when you hold the delete character key. There is a lot of delay between the moment you start holding the key and the moment you see your whole sentence removed. When it reaches that point and I stop holding the erase key, the keyboard continues erasing a lot of characters. It also feels very laggy.

The previous version was perfect on that point (when I stopped holding the key, the keyboard would immediately stop erasing characters + I could see all characters getting removed one by one).

Link to the app: https://f-droid.org/packages/net.gsantner.markor/

Although I'm writing to complain, I also wanted to thank you very much for the dedication you put into this project, keep up the good work ;)

**Reply by patrickgold** [reply]
Thanks for the feedback!

Installed the Markor app from F-Droid and did not have any issues at all with the delete key. May I ask which Android version/ROM and device you are on? There could be an edge case with the input connection I missed which causes the delay.

**Reply by Humorist2601** [reply]
It works for me aswell on Android 12.1 (ArrowOS) using both Delete Character and Delete Word

**Reply by papjul** [reply]
I checked with a new TODO file and it worked fine indeed. I filled up this file with 5000 characters and it still worked as if it were new. I doubled to 10 000 characters (which is the number of characters I have in my real TODO file) and it started having the issues.

I could not reproduce the issue with the 10 000 characters copied to a regular text app (with no markup then) so I guess it has to do with the rich text editor which is probably not optimized for so many characters being a part of the problem.

I know the best thing I can do is clear my todo list but it used to work fine with previous Florisboard version so I wanted to report it in case it happens in other similar editors. To be perfectly honest, Florisboard previous version worked even better than the native Android keyboard (which also has some issues erasing characters but not as problematic as Florisboard new version).

I'm using Lineage 19.1 on an XZ1 Compact.

**Reply by patrickgold** [reply]
Ok so I tested Markor a bit more thoroughly (in a TODO list format with just shy of 20k characters) and I can reproduce your described issue, however this is not FlorisBoard's fault:

Every time Markor re-evaluates how to color its text it delays input from a keyboard until its re-evaluation is done. This means regardless if you press backspace or try to input a character it will lag if you input it right when it tries to determine the color again. You can also exactly see when the keyboard will lag, because during re-evaluation the input cursor disappears and does not blink.

I tested this with FlorisBoard v0.3.15, v0.3.16-beta01, AnySoftKeyboard and Gboard on Android 11, all producing roughly the same lag. This is clearly an issue with Markor and I cannot fix an input lag which is not caused by my keyboard.

What confuses me is that you mention it worked better for you in 0.3.15. I tested it and it lags the same with the old logic, if not even worse because of the asynchronous logic still in there. Do you by any chance have a screen recording which shows the v0.3.15 (or a previous beta) in Markor (TODO list with colors)?

One must say though that regular markdown with also ~20k characters works perfectly fine in my tests with Markor, so its clearly the rich text formatting that's the issue here.

**Reply by papjul** [reply]
Thank you for your technical analysis!

Here is the first example with 0.3.15-rc02
https://user-images.githubusercontent.com/1449159/167510631-66ee1bfe-ae8b-4d93-985b-9787d9a4215b.mp4

And now with 0.3.16 (I had to hold off the backspace key when it reached 2022-05 because it was way too slow, but it still removed two more lines that I didn't ask to remove):
https://user-images.githubusercontent.com/1449159/167510645-40414253-597b-422c-910d-e1565e9bc246.mp4

In first video, it seems to me it correctly stop erasing approximately when I hold off my finger from the backspace key, so I'm not sure to understand Markor's responsibility here.

Let me know your findings and if I need to report a bug to Markor ;)

**Reply by patrickgold** [reply]
Thanks for the provided screen recordings! The 0.3.15 screen recording roughly matches with my one, it has a small lag, but generally it is correct.

The 0.3.16 screen recording however is completely different and I don't really know how this even happens. Could you please do another screen recording, this time with some debug info enabled?  
Before doing the screen recording, go to FlorisBoard Settings > Devtools > (Enable if not already) > Show input state overlay. This will overlay important debug info about the current editor state and is valuable for me to hopefully pinpoint the issue.

Thanks in advance!

**Reply by papjul** [reply]
Here it is, I released the backspace key when cursor was at the end of "2022-05-10 ":
https://user-images.githubusercontent.com/1449159/167570843-64e24af7-c5e9-4fe9-8aed-b6ba8d52f70e.mp4

**Reply by patrickgold** [reply]
Ok the only thing I can imagine is that the batch edits overlap each other and Markor reacts badly to that? I honestly am unsure what causes this. Will investigate later into this more.

**Reply by papjul** [reply]
The issue is fixed either in beta 2 or 3. I cannot reproduce with beta 3.

**Comment by HikaruIchijyo** [comment]
Somehow after 0.3.16 beta even if clipboard is empty/cleared sometimes last copied msg shows in toolbar is it bug/know?

Reproduce:
Copy any text
Open clipboard
Clear all
Close keyboard when u open again boom last copied shows. When u open clipbpard it is empty close again & when u open again it will appear again.


On miui 13 redmi note 10s

**Reply by patrickgold** [reply]
Commit https://github.com/florisboard/florisboard/pull/1832/commits/60221743f6a7721bb006cc946fcf19ddfb308d5c in PR #1832 fixes this issue. It was caused due to a missing update logic for the Smartbar suggestions UI. Thanks for reporting, this will be parts of the 0.3.16-beta02 release!

**Reply by HikaruIchijyo** [reply]
> Commit https://github.com/florisboard/florisboard/pull/1832/commits/60221743f6a7721bb006cc946fcf19ddfb308d5c in PR #1832 fixes this issue. It was caused due to a missing update logic for the Smartbar suggestions UI. Thanks for reporting, this will be parts of the 0.3.16-beta02 release!

Thanks

**Reply by HikaruIchijyo** [reply]
> Commit https://github.com/florisboard/florisboard/pull/1832/commits/60221743f6a7721bb006cc946fcf19ddfb308d5c in PR #1832 fixes this issue. It was caused due to a missing update logic for the Smartbar suggestions UI. Thanks for reporting, this will be parts of the 0.3.16-beta02 release!

Somehow it is still present can u chk & confirm?

**Reply by patrickgold** [reply]
Ah I see now you meant something else, my bad. This behavior isn't new though or a bug, clearing the history didn't clear the primary clip in 0.3.15 or 0.3.14 either. This is because they are decoupled, the history is a database and the primary clip just an object holding the clip data.

I can look into making a toggle for this in the clipboard prefs, like there's already one for the other direction ("Clear primary clip affects history"). I think your request is something along #811, am I right?

**Reply by HikaruIchijyo** [reply]
> I can look into making a toggle for this in the clipboard prefs, like there's already one for the other direction ("Clear primary clip affects history"). I think your request is something along #811, am I right?

I guess soo
Well previously it wont go away i mean the last copied thing no matter the time.

Currently after .02 beta when u clear history it persists fr sometimes & then automatically disappears. So it is kinda something like when u have something cached & after sometimes thats get cleared.

**Comment by ShareASmile** [comment]
I would Request support for devices on Android 6 Api Level 23.
..maybe an extra APK for devices still on this API level will do if it is maintained for a bit longer time. Thank You
Edit:  Sorry for posting this request here i realised this mistake later as this discussion  thread is specific to 0.3.16 input logic and editor but please do consider request for Android 6 support.

**Comment by VesperLlama** [comment]
In 0.3.16, the keyboard sometimes opens with the characters in lowercase even if there is no text in the input field. It is easy to reproduce this in messaging apps like Telegram and WhatsApp but I have also seen it in other apps too but it's rarer.

To reproduce - 
1. Send a message in a messaging app like Telegram or WhatsApp.
2. Close the keyboard after sending the message.
3. Open the keyboard again. The characters will now be in lowercase instead of uppercase.

Reopening the keyboard doesn't fix it. It gets fixed after pressing any keyboard button and deleting it.

**Reply by patrickgold** [reply]
Could you please confirm that this bug is still present in 0.3.16-beta02? I can somehow not really reproduce this.

**Reply by VesperLlama** [reply]
Yes, it's still there in beta02. I even did a reinstall and it was still there with the default settings. I have attached a recording of the bug. I am on Android 11. Is there a way to get logs for this?

https://user-images.githubusercontent.com/74611933/168121476-775a364c-8c3c-42af-908d-3f8e1dc3d165.mp4

**Reply by patrickgold** [reply]
Ah I see now you are right, can reproduce this in WhatsApp and Signal. WIll try to solve this for the next beta release!

**Reply by VesperLlama** [reply]
This has been fixed in beta04. Thanks!

**Comment by pierstitus** [comment]
swiping up from the spacebar now opens the popups on letters and types them when the gesture is not recognized. The spacebar up gesture was already inconsistent, #1546, but didn't type unintended characters.

**Comment by billie80** [comment]
for some input fields the keyboard does not start up with capital letters. Up to now I encounter it only in the web browser (Fennec from F-Droid here). Proof ist the first letter. 😄
Does not happen with the current stable version.

**Reply by billie80** [reply]
I guess this is issue https://github.com/florisboard/florisboard/issues/1861. Let's see if beta04 fixes it.

Edit: Looks like it did. 👍 Just the settings are gone as usual. Thought there were fixes for this too. 😀

**Reply by patrickgold** [reply]
> Looks like it did. 👍

Yeah I forgot to link in this thread, #1861 was the tracking issues for this. Great to hear that it is fixed for you too!

> Just the settings are gone as usual.

Could you provide some details for this in #1726? I tried to fix this in the beta04 but it seems lime this fix does not help everyone.

**Comment by magnus-ISU** [comment]
Creating themes seems to be broken.

When glide typing, I wasn't getting any spaces between words.

**Comment by jonas-w** [comment]
Cosmetic bug:

In the settings Keyboard -> Key spacing

The values only can be in 0.5 steps but while sliding the slider it shows an exact value with 7 decimal places which is a bit weird. A slider with 0.5 steps would be better here

**Reply by patrickgold** [reply]
I am aware, however currently I don't know how to fix this, as the slider API for float values is somewhat limited.

**Comment by HikaruIchijyo** [comment]
W8 in primary acyion smart bar hisyory icon is removed or something?

**Reply by patrickgold** [reply]
> W8 in primary acyion smart bar hisyory icon is removed or something?

Nah it is still there, although it has moved one right and thus won't display on smaller screens. This was needed because I added the voice key to the primary actions. If you don't see the history icon then you can either use the icon in the secondary row or by long-pressing the enter key, both ways will get you to the same screen. Re-ordering of the Smartbar actions is something which is needed and I am aware of this, however it will still take some time until this gets implemented, mainly as I will soon start to focus on word suggestions.

**Reply by HikaruIchijyo** [reply]
> > W8 in primary acyion smart bar hisyory icon is removed or something?
> 
> Nah it is still there, although it has moved one right and thus won't display on smaller screens. This was needed because I added the voice key to the primary actions. If you don't see the history icon then you can either use the icon in the secondary row or by long-pressing the enter key, both ways will get you to the same screen. Re-ordering of the Smartbar actions is something which is needed and I am aware of this, however it will still take some time until this gets implemented, mainly as I will soon start to focus on word suggestions.

Thanks 
Sure np

**Comment by tsiflimagas** [comment]
One thing I've noticed since some time is that cpu usage is kinda high when typing. The way I noticed this is by a (ingenious, lol) kind of profiling I do. Open a terminal app and run `top -d 0.5 -p $(pgrep -f <package name>)`, then start randomly typing. There's probably a much better way to do it in Android Studio. Typing in the terminal doesn't raise cpu usage too much (about 40-50%, which still is high), as it seems, since there's no spell checking being performed. But typing in a field which enables spell checking, seems pretty expensive, using about 80% on average, and going over 100% at times (I do this using split screen😑). I hope you have some ideas or can find out what's maybe not quite right, without having to reintroduce any regressions of the ones you fixed in #1822.

**Reply by patrickgold** [reply]
That's really difficult to fix, because CPU usage was always a problem within this keyboard (sometimes better sometimes worse) and the 0.3.16 logic keeps the content and keyboard state always updated, sadly at the cost of CPU. I may need to investigate the observer logic to ensure I didn't accidentally introduced observer loops. Also I need to check if the `android.icu` platform tools must be used with care and as little as possible if they are CPU heavy.

> There's probably a much better way to do it in Android Studio

There is (profiler with call stack view), however I am not that well trained in using this tool yet tbh, must still learn a lot to get better at utilizing it correctly.

**Reply by patrickgold** [reply]
See #1894 for a few efforts of me trying to reduce high CPU usage. Currently the PR has one fix addressing high CPU usage in the input feedback controller due to repeated expensive system calls that are avoidable. In the profiler I could verify that this specific source of lag is now gone (and the system CPU usage now does not jump as high as previously when pressing keys), however overall to the eye no real improvement is felt (at least for me).

I will continue with some experiments in the profiler tomorrow (that tool is actually pretty good once I study it a bit more how it works), hopefully achieving some other minor improvements.

**Reply by tsiflimagas** [reply]
Great, thank you for working on this! I'll be also testing on my professional profiler😁 I think a bit of decrease is already visible in cpu usage!

**Reply by patrickgold** [reply]
For now I am done with my performance improvements in #1894. It isn't a major leap in CPU usage reduction in the end, however I fixed some small issues especially surrounding the logic of selection update and touch logic, which are performance-critical parts and every small fix can lead to some improvement.

Still some major issues I noticed in the profiler do remain:
1. After a clean cold start of the keyboard, the first few keys require a lot of CPU power (up to 100%). For powerful devices this is not noticeable without monitoring the CPU usage, for weaker devices this produces lag (this is also reported by some users, so it seems that they are on weaker devices). I don't really know why this occurs, and haven't really found the source in the profiler either. Definitely room to improve here in the future.
2. Calls to the input connection are very expensive performance-wise, and I think I need to reduce or optimize the calls. I did that with the auto-capitalization (which is now evaluated directly within the keyboard and basically removes the IPC round-trip cost of calculating the current caps mode).
3. Calls to native ICU methods (especially when it comes to looking up special values in ICU tables) also don't come cheap and cause a lot of thread-blocking. I tried to cache results here (e.g. space bar language name was a culprit), however I am sure there are other points in the app where this is also a performance issue.

All in all I think #1894 is an improvement (and I did a lot of testing to ensure I didn't break anything accidentally). If possible, could you just check the performance of the app and also if the changed features still behave as they should in 0.3.16-rc01 (auto-capitalization has some fixes so it behaves correctly now as supposed to beta04)? Having a second pair of eyes look over that really helps to ensure I can then safely release 0.3.16. Thanks a lot in advance!

**Reply by tsiflimagas** [reply]
As you said, it's not a radical change, but I think measurements are a tad better indeed. On usage, i'm not sure if I can see the difference, as it was good already. Perhaps it's a little more responsive at times, especially when there are other things running too, but i'm usually catious about such empirical observations.
As for stability, I didn't find anything to be broken. Auto capitalization works as expected in different cases. Testing was done on release build btw, I often make local builds (I mean, it wouldn't make sense to test debug builds for performance).

> After a clean cold start of the keyboard, the first few keys require a lot of CPU power (up to 100%)

I noticed that too in my testing. My cpu can't be considered that powerful, thus this is quite obvious on monitoring (maybe a little bit on usage too). Idk if JC could be to blame here.
Another thing I just noticed (not newly introduced), is that using swipe to delete, swipe to move cursor and text selection by holding shift key is weirdly heavy on cpu. It steadily stays over 100% till the gesture is completed, which seems like there's some constant and perhaps unnecessary activity going on concurrently. Well, maybe not, just a speculation.
Good to see that you have more observations and thoughts already anyway. It means that we can probably expect further improvements in this area in the future.

**Comment by ghost** [comment]
Florisboard 3.16 beta 04
Android 12
Single key input appears to 'grab' the characters faster.  Swipe has much more misspellings.  <These are both related?  
I can feel multiple initiations of vibrations when I start a word with swipe.  There should only be one instance.  Its difficult to work with, like an echo on the phone when you're speaking.  I don't know if this is a thing, and without knowing anything about what has been tweaked, i'm wondering if the sensitivity of character recognition is so high that the phone's haptic vibration feedback is interfering with swipe word recognition. ?  In other words the vibrations feeding into the touch sensor, causing it to 'skip' and genetate multiple initiations. 

I have all of these settings ON: 
Enable haptic
Trigger vib dir
Key press vib
Key repeated action vib
Gesture moving swipe vib

General feedback::  Why so many languages?  Why so many vibration settings?  <These make me wonder if resources are spent on things of low importance, whereas word recognition is of high importance, and is still poor.  

I typed this out using this keyboard, and it took me so long, due to me needing to correct many of these words.

**Reply by tsiflimagas** [reply]
> Single key input appears to 'grab' the characters faster. Swipe has much more misspellings. <These are both related?

Do you mean "more accurately" instead of "faster"? Else, I can't quite understand the sentence. But if that's the case, then of course it's like that, and I think it's obvious why. When you tap on a key, it's totally precise (expect if you happen to tap inaccurately, but tapping isn't that hard), but when swiping, you can't be totally precise on every character your finger passes by, in fact you may be very approximate, so the algorithm has to suppose that you meant to pass by adjacent characters all the time, and attempt to make predictions accounting for this. So, I don't quite get what you mean by whether these two are related, since they're two different things. Perhaps I got the question wrong. 

> I can feel multiple initiations of vibrations when I start a word with swipe

Yeah, just tried and I can confirm there are two consecutive vibrations at the start of the gesture. I guess that's a bug indeed.

> Why so many languages? Why so many vibration settings? <These make me wonder if resources are spent on things of low importance, whereas word recognition is of high importance, and is still poor.

By languages do you mean keyboard layouts, or settings translations? I guess the former. Well, if being able to at least type on multiple languages, even without the ease of word suggestions etc, isn't considered the most major part of a keyboard, then I don't know what is. But, even then, it isn't even so resource demanding. It's literally easy to add a new layout (well, some are a little or even much harder indeed), and many, if not most, of them are added by contributors. Also, having multiple options for vibration control or any other setting, perhaps isn't even that time consuming to add. Not to underestimate the work done there, of course, I'm rather saying this in the sense that it didn't "steal" time that would have been spent on something big, which would have required exclusive work for a much bigger chunk of time anyway. It's just that when working on a new feature, it seems to me that Patrick wants to make it as good as it gets, instead or iterating again and again, so time is spent worthily, regardless.
As for word recognition precision when glide typing, I can tell that there haven't been many complaints, if any at all. From my personal experience too, it's pretty good. I've used Gboard's glide typing also, and I can't tell these two are far, it's just that glide typing isn't meant for high preciseness by design, except if you're really good at it, at least as I see it. Thus, there haven't been any reason to work on improving it. Though, some introduced issues were resolved in #1844, so some job was done on it indeed. 
Furthermore, the code for the algorithm was taken from a pull request of another dev who contributed it to AnySoftKeyboard, and was implemented to FlorisBoard by @X-yl. So, going through that complex algorithm someone else created and improving it further, would probably be a hell lot of job, perhaps nearly impossible, while there's a hell lot of things to do aside.
I understand ones concerns about whether the planning of the project is good and time is well spent on it, but I think that it needs quite some more to make a judgement on it than just a glimpse.

Note: I'm not a dev here or in general, so anything I write can be faulted. Pardon me if I got redundantly verbose somewhere 😄

**Reply by MikeCheque** [reply]
android 12
0.3.16-beta04 (84)

no key pad vibrations are felt on keypress despite all settings enabled
![Jun-8,-2022-3-44-10-PM](https://user-images.githubusercontent.com/45575396/172704789-655a59aa-4821-4ff8-af93-fee480cff18a.jpg)

**Comment by trizzosk** [comment]
Ehm... Android 12 (GrapheneOS) on Pixel 6 Pro: with this release, keyboard is laggy, popup after key press only shows when I write very slowly... I have to use amother keyboard. I used this because ot was for me the most precious, fast and reliable keboard. Now I cannot use it.

**Reply by tsiflimagas** [reply]
Which release exactly? You didn't specify a version.

**Reply by trizzosk** [reply]
I was directly responding from f-droid... 0.3.16 RC01.

**Reply by tsiflimagas** [reply]
Was previous beta fine performance-wise? Version 0.3.16-rc01 got a few more performance improvements, so it doesn't quite make sense for it to be worse, except if something got broken accidentally.

**Reply by trizzosk** [reply]
I just switched back to 0.3.15 (I guess the latest stable release) and I don't have any issues. Yeah maybe something accidental. I saw and read release notes. Well, I will wait until RC02 or newer will be released and try it after.

**Reply by tsiflimagas** [reply]
I think the next one is going to be 0.3.16 stable, unless the need for another release candidate comes up. Had you tried any other of the 0.3.16 betas previously though? I want to understand if rc01 is the only version to have this issue for you.

**Reply by patrickgold** [reply]
Seems like something accidentally broke. Looking at the diff I don't see too much reason though what could have broken so horribly that you experience such lag (also as I haven't received such report from anyone else looking only at the 0.3.16-rc01).

If possible could you back up the FlorisBoard settings (FlorisBoard > Advanced > Back up), then delete the cache and app data from the FlorisBoard Beta app, force kill it (is normally the button next to delete data in the system prefs UI) and set up the Beta again? Is the experience then smooth, and if so does it change again once you restore your backup?

**Comment by 2cliptrip** [comment]
> This discussion serves as a collection for all feedback/issues surrounding the new input connection logic introduced in #1822 and [0.3.16-beta01](https://github.com/florisboard/florisboard/releases/tag/v0.3.16-beta01).
> 
> Please post your feedback here if it concerns the following:
> 
> * General feedback to the overhaul: Is input now faster/laggier/the same as before? Did the input experience change for the better/the worse/stayed the same?
> * Input into specific or any apps does not work correctly or at all
> * Input randomly overwrites stuff
> * Selection/cursor movement is messed up
> * Delete character/words does not work
> * Other issues in this area
> 
> If you have an issue in this area which is new since 0.3.15, please make sure to post it here (also specify the app and Android version you are on) and I will try to fix them in the following beta releases.
> 
> Thanks a lot!
> 
> See PR #1832, which addresses feedback gathered from this thread and other issues

GReat key board and alot of improvement. I would like inline edit for thr clipboard make it and autospace but overalll hats off

**Comment by Mehdiibeigii** [comment]
Tanks
